rust/lance-table/src/io/manifest.rs

```diff
@@ -253,4 +256,41 @@ mod test {
         test_roundtrip_manifest(1000, 100_000).await;
         test_roundtrip_manifest(1000, 1000).await;
     }
+
+    #[tokio::test]
+    async fn test_update_schema_metadata() {
+        let store = ObjectStore::memory();
+        let path = Path::from("/update_schema_metadata");
+
+        let arrow_schema = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+            "i",
+            DataType::Int32,
+            false,
+        )]));
+        let schema = Schema::try_from(arrow_schema.as_ref()).unwrap();
+        let mut file_writer = FileWriter::<ManifestDescribing>::try_new(
+            &store,
+            &path,
+            schema.clone(),
+            &Default::default(),
+        )
+        .await
+        .unwrap();
+
+        let array = Int32Array::from_iter_values(0..10);
+        let batch =
+            RecordBatch::try_new(arrow_schema.clone(), vec![Arc::new(array.clone())]).unwrap();
+        file_writer.write(&[batch.clone()]).await.unwrap();
+        file_writer
+            .schema
+            .metadata
+            .insert(String::from("lance:extra"), String::from("for_test"));
```

Comment:
Why can't the metadata be passed in the beginning?

Reply:
Because these metadata might not known when open a file writer. 
For example, the offset of certain blocks during the writes, or the range of batch IDs of particular data. 
i.e. if we open a `FileWriter` and write HNSW layers as batches, we can collect the batch IDs of each layer, where some large layer could have many batches,  and track them in the schema metadata.

Reply:
I worry this is exposing too much of the internals.

Instead of batch ids, we should probably keep track of the row ranges. Weston would like to do away with the idea of batches in files for V2.

Could we just have an API `.add_metadata(key, value)` that lets you add additional metadata as long as the writer hasn't been finished?

Reply:
Yeah, the row groups are going away, I think you'd just use row ranges instead of batch IDs in the future.  I don't think there is a problem with writing the data in batches.  But what do you hope to do with these batch IDs?

That being said, maybe metadata should get passed into the `FileWriter` as part of the finish function?  I think that would make sense.

E.g. maybe you want to build up a histogram of some column while you are writing the file and then store that histogram as metadata.

Reply:
Ok, i added a `finish_with_metadata()` method.